src/Pos/Txp/Toil/Utxo/Functions.hs

```diff
@@ -1,62 +1,232 @@
+                      i sumDist txOutValue)
+           , (allDistinct (map fst d :: [StakeholderId]),
+              sformat ("output #"%int%"'s distribution "%
+                       "has duplicated addresses")
+                      i)
+           , (all (> mkCoin 0) (map snd d),
+              sformat ("output #"%int%"'s distribution "%
+                       "assigns 0 coins to some addresses")
+                      i)
+           ]
+
+verifyInputs :: VTxContext
+             -> NonEmpty (TxIn, TxOutAux)
+             -> TxAux
+             -> VerificationRes
+verifyInputs VTxContext {..} resolvedInputs (view txOutputs -> outs, witnesses, distrs) =
+    verifyGeneric . concat $
+    zipWith3 inputPredicates [0 ..] (toList resolvedInputs) (toList witnesses)
+  where
+    txOutHash = hash outs
+    distrsHash = hash distrs
+    inputPredicates
+        :: Word32           -- ^ Input index
+        -> (TxIn, TxOutAux) -- ^ Input and corresponding output data
+        -> TxInWitness
+        -> [(Bool, Text)]
+    inputPredicates i (txIn@TxIn{..}, toa@(TxOutAux txOut@TxOut{..} _)) witness =
+        [ ( checkAddrHash txOutAddress witness
+          , sformat ("input #"%int%"'s witness doesn't match address "%
+                     "of corresponding output:\n"%
+                     "  input: "%build%"\n"%
+                     "  output spent by this input: "%build%"\n"%
+                     "  address details: "%addressDetailedF%"\n"%
+                     "  witness: "%build)
+                i txIn txOut txOutAddress witness
+          )
+        , case validateTxIn txIn toa witness of
+              Right _ -> (True, panic "can't happen")
+              Left err -> (False, sformat
+                  ("input #"%int%" isn't validated by its witness:\n"%
+                   "  reason: "%build%"\n"%
+                   "  input: "%build%"\n"%
+                   "  output spent by this input: "%build%"\n"%
+                   "  witness: "%build)
+                  i err txIn txOut witness)
+        ]
+
+    checkAddrHash addr wit = case wit of
+        PkWitness{..}          -> checkPubKeyAddress twKey addr
+        ScriptWitness{..}      -> checkScriptAddress twValidator addr
+        RedeemWitness{..}      -> checkRedeemAddress twRedeemKey addr
+        UnknownWitnessType t _ -> checkUnknownAddressType t addr
+
+    -- third argument here is local context, can be used for scripts
+    validateTxIn :: TxIn -> TxOutAux -> TxInWitness -> Either String ()
+    validateTxIn TxIn {..} _ PkWitness {..} =
+        unless (checkSig twKey (txInHash, txInIndex, txOutHash, distrsHash) twSig) $
+        Left "signature check failed"
+    validateTxIn TxIn{..} _txOutAux ScriptWitness{..}
+        | scrVersion twValidator /= scrVersion twRedeemer =
+            Left "validator and redeemer have different versions"
+        | not (isKnownScriptVersion (scrVersion twValidator)) =
+            Right ()
```

Comment:
I wanted to write `pass` here, but I thought that @neongreen would say that `Right ()` is better because you know exact type so I decided not to touch it.
P. S. Most of comments to this issue are not about code I wrote (I only moved it).